### ISSUE TITLE
[RKE1] kube-controller-manager / kube-scheduler HTTPs ports cannot perform authn / authz by default

### ISSUE THREAD

**Issue by aiyengar2** [rancher/rancher#35751]
**Rancher Server Setup**
- Installation option (Docker install/Helm Chart):
   - If Helm Chart, Kubernetes Cluster and version (RKE1, RKE2, k3s, EKS, etc): RKE1

**Information about the Cluster**
- Kubernetes version: All (as of v1.21.6-rancher1)
- Cluster Type (Local/Downstream): either
   - If downstream, what type of cluster? (Custom/Imported or specify provider for Hosted/Infrastructure Provider):

**Describe the bug**

kube-controller-manager / kube-scheduler's HTTPS ports in any RKE1 cluster today with default settings cannot perform authn / authz requests to verify any requests coming to the HTTPs ports (except some paths like healthz, which bypass these checks) since a kubeconfig is not provided to be able to perform these requests. As a result, the HTTPS ports are effectively unusable unless a user manually overrides the necessary arguments on installing / upgrading RKE1 (as any user who tries to hit the HTTPs endpoint causes the component to reach out to the API server to check their credentials, which the component cannot do without a valid kubeconfig to use to check those credentials).

While this was not previously an issue for RKE1 since we would communicate over the HTTP ports, in k8s 1.22 / 1.23 the HTTP ports are being removed by default (and disabled in k8s 1.24), so RKE needs to fix the ability to make calls to the HTTPS ports for these components to ensure other components like Monitoring can still make requests to those endpoints.

**To Reproduce**
 

- Spin up an RKE1 cluster
- Grab an auth token: for example, use the following one-line for the default token`TOKEN=$(kubectl get secrets -o jsonpath="{.items[?(@.metadata.annotations['kubernetes\.io/service-account\.name']=='default')].data.token}"|base64 -d)`
- Log onto a controlplane host and try to run `curl https://localhost:10257/metrics -k -H "Accept: application/json" -H "Authorization: Bearer ${TOKEN}"`

**Result**
No matter what token you use, you will always see a 403 to query the endpoint due to the fact that you are an **anonymous** user.

**Expected Result**

You should either see a 403 that shows that **you** (not an anonymous user) don't have access to controller-manager or correctly receive the metrics if you are using the token of a service account who has permissions to query `/metrics`, e.g. an entity bound to a ClusterRole that has these permissions: https://github.com/rancher/charts/blob/73f1cd647751650190644b23d0e41221bdfed644/charts/rancher-pushprox/templates/pushprox-clients-rbac.yaml#L14-L15.

**Screenshots**

**Additional context**

Before k8s `1.22` for kube-controller-manager (changed in https://github.com/kubernetes/kubernetes/pull/96216) and before k8s `1.23` for kube-scheduler (changed in https://github.com/kubernetes/kubernetes/pull/96345), both k8s components would be exposed in two ways:
- a **HTTPS** port (10257 for controller-manger, 10259 for scheduler)
- an **HTTP** port (10252 for controller-manager, 10251 for scheduler).

Generally, communication between Rancher components and these services would be transmitted over the HTTP port, e.g. rancher-monitoring would scrape metrics from these HTTP ports, as seen in https://github.com/rancher/charts/blob/73f1cd647751650190644b23d0e41221bdfed644/charts/rancher-monitoring/values.yaml#L27-L55.

However, following k8s 1.22 / 1.23, **only the HTTPS port will be enabled for these components**. 

However, the issue today is that RKE1 does not pass in the necessary arguments `--authentication-kubeconfig` and `--authorization-kubeconfig` to both components to give the components a valid kubeconfig to make authn / authz checks with the API server; as a result, any request made to the secure port that requires any form of authn / authz (e.g. most calls except healthz calls) is automatically denied due to the fact that the component cannot figure out who the user is and whether they are allowed to access that path on the component.

**Workaround**

For clusters <1.21, just use the HTTP port (10252, 10251).

For clusters 1.22+, this is a blocker and the HTTP port is disabled by default. Users could enable the HTTP port by overriding arguments.

For clusters 1.24+, this is a blocker since the HTTP port is disabled with no ability to re-enable.

**Comment by kinarashah** [comment]
We should also look into updating these args for other k8s versions, https://github.com/rancher/rke/pull/2763#issuecomment-985945268

**Comment by aiyengar2** [comment]
RKE2 already resolved this issue in 1.22 in https://github.com/k3s-io/k3s/commit/dc556cbb727a86c5b11811cf5c0b3fb2780b3ae8#diff-e0ce1a7f3f76d7bfda0ffdae6e0226aa63b3009c4c4a770c321876aecad0a6c1, as found in comments in the discussion on https://github.com/rancher/rke2/issues/1903.

**Comment by aiyengar2** [comment]
Currently, the fix being performed for this issue is to add the required arguments to get the HTTPS port working to RKE1 1.22 **only**.

Do we need to backport this fix to RKE1 1.17, 1.18, 1.19, 1.20, and 1.21 release lines as well or are users expected to follow the workaround of using the HTTP port or setting the arguments manually or upgrading to 1.22 if they want to be able to use the HTTPS port as well?

cc: @deniseschannon @Jono-SUSE-Rancher

**Comment by cbron** [comment]
We are going to leave in 2.6.3 for RKE1 1.22, too confusing to revert now.

**Comment by annablender** [comment]
Validated on DO RKE1 downstream cluster with k8s 1.22.

Steps
1. Run Rancher 2.6-head on local k3d cluster
2. Create single-node DO RKE1 downstream cluster
3. Get token - run TOKEN=$(kubectl get secrets -o jsonpath="{.items[?(@.metadata.annotations['kubernetes\.io/service-account\.name']=='default')].data.token}"|base64 -d) on localhost
4. Login into DO control-plane node
5. Run curl https://localhost:10257/metrics -k -H "Accept: application/json" -H "Authorization: Bearer ${TOKEN}"

Result

```
root@ablendermann1:~# curl https://localhost:10257/metrics -k -H "Accept: application/json" -H "Authorization: Bearer ${TOKEN}"
{
  "kind": "Status",
  "apiVersion": "v1",
  "metadata": {
    
  },
  "status": "Failure",
  "message": "forbidden: User \"system:anonymous\" cannot get path \"/metrics\"",
  "reason": "Forbidden",
  "details": {
    
  },
  "code": 403
}
```

kube-controller-manager is able to make a HTTPS request to the metrics endpoint with authorization and RKE1 was able to identify the user (as not having permissions) because it has the kubeconfig credentials. This will _only_ be available for RKE 1.22 in 2.6.3. QA testing in progress.

**Comment by annablender** [comment]
Validation that RKE args work is done. Issue ok to close.
